engine/class_modules/sc_demon_hunter.cpp

```diff
@@ -4640,6 +4768,10 @@ struct soul_cleave_t : public demon_hunter_attack_t
 
   void execute() override
   {
+    soul_cleave_state_t* soul_cleave_state = debug_cast<soul_cleave_state_t*>(execute_action->get_state());
+    soul_cleave_state->t29_vengeance_4pc_proc = rng().roll( p()->set_bonuses.t29_vengeance_4pc->effectN( 2 ).percent() );
+    execute_action->snapshot_state(soul_cleave_state, result_amount_type::DMG_DIRECT);
```

Comment:
This will actually leak memory since the state isn't assigned anywhere due to this just being a normal execute. There's ways you could add this into the execute_state but actually think the cleanest way to handle this (and could do the same thing for Soul Bomb) is just inject the t29_vengeance_4pc_proc set/rng into the snapshot_state() function itself.

```
void snapshot_state( action_state_t* state, result_amount_type rt ) override
 {
     demon_hunter_attack_t::snapshot_state( state, rt );

     if( p()->set_bonuses.t29_vengeance_4pc->ok() && rng().roll( p()->set_bonuses.t29_vengeance_4pc->effectN( 2 ).percent() )
       debug_cast<soul_cleave_state_t*>(state)->t29_vengeance_4pc_proc = true
}
```

Or something along those lines.